src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/LengthBucketsFrozenDictionary.cs

```diff
@@ -1,29 +1,27 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Buffers;
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.Runtime.CompilerServices;
-using System.Runtime.InteropServices;
 
 namespace System.Collections.Frozen
 {
     /// <summary>Provides a frozen dictionary implementation where strings are grouped by their lengths.</summary>
     internal sealed class LengthBucketsFrozenDictionary<TValue> : FrozenDictionary<string, TValue>
     {
-        /// <summary>Allowed ratio between buckets with values and total buckets.  Under this ratio, this implementation won't be used due to too much wasted space.</summary>
-        private const double EmptyLengthsRatio = 0.2;
```

Comment:
I don't understand why it's ok to do away with this concept.  If I have one string of length 1_000_000 and one string of length 1, won't I now end up allocating an array of 5 million elements?

Reply:
> This case is the first thing we check and I've left it untouched:

No, that's different.  The check that's there is seeing whether it can be easily predetermined whether any bucket might have too many elements in it.  In contrast, the EmptyLengthsRatio you deleted was used to determine whether the lengths were too spread out such that we'd be wasting a ton of space with holes in the array.

Reply:
> I am going to benchmark whether it's beneficial to remove this check or not and get back to you.

For MaxPerLength you mean? For a given bucket, the implementation is doing a linear scan of every element.  If 1000 strings were to end up in the same bucket and every look up involved comparing up to 1000 strings, it would absolutely be worse than an O(1) dictionary lookup.  So there is going to be some threshold where we no longer want to use this strategy.  It's certainly possible, especially with the other changes, that MaxPerLength can be increased, but it can't become infinite.